datumaro/components/visualizer.py

```diff
@@ -0,0 +1,129 @@
+# Copyright (C) 2022 Intel Corporation
+#
+# SPDX-License-Identifier: MIT
+
+import warnings
+from typing import Iterable, List, Optional, Tuple, Union
+
+import matplotlib.patches as patches
+import matplotlib.pyplot as plt
+import numpy as np
+from matplotlib.axes import Axes
+from matplotlib.figure import Figure
+
+from datumaro.components.annotation import Annotation, AnnotationType, Bbox, LabelCategories
+from datumaro.components.dataset import IDataset
+from datumaro.components.extractor import DatasetItem
+
+DEFAULT_COLOR_CYCLES: List[str] = [
+    "#1f77b4",
+    "#ff7f0e",
+    "#2ca02c",
+    "#d62728",
+    "#9467bd",
+    "#8c564b",
+    "#e377c2",
+    "#7f7f7f",
+    "#bcbd22",
+    "#17becf",
+]
+
+TEXT_Y_OFFSET = 1.0
+
+
+class Visualizer:
+    def __init__(
+        self,
+        dataset: IDataset,
+        ignored_types: Optional[Iterable[AnnotationType]] = None,
+        figsize: Tuple[float, float] = (8, 6),
+        color_cycles: Optional[List[str]] = None,
+        bbox_linewidth: float = 1.0,
+        text_y_offset: float = 1.5,
+    ) -> None:
+        self.dataset = dataset
+        self.figsize = figsize
+        self.ignored_types = set(ignored_types) if ignored_types is not None else set()
+        self.color_cycles = color_cycles if color_cycles is not None else DEFAULT_COLOR_CYCLES
+        self.bbox_linewidth = bbox_linewidth
+        self.text_y_offset = text_y_offset
+
+    def _sort_by_z_order(self, annotations: List[Annotation]) -> List[Annotation]:
+        def _sort_key(ann: Annotation):
+            z_order = getattr(ann, "z_order", -1)
+            return z_order
+
+        return sorted(annotations, key=_sort_key)
+
+    def vis_gallery(
+        self,
+        ids: List[Union[str, DatasetItem]],
+        nrows: int,
+        ncols: int,
+        subset: Optional[str] = None,
+    ) -> Figure:
+        assert nrows > 0, "nrows should be a positive integer."
```

Comment:
I think that halting the program is not a good idea since the visualizer is 'only for showing' not a core functional logic.
How about just printing error messages and showing empty images when something goes wrong instead of assertion?

Reply:
My thought is that this API is mainly target for Jupyter Notebook users. So, if there is a problem, their executions will be failed only at a cell scope. I think it will be okay and giving exception can make users more cautious than printing messages. If this API will be used in Python programs, I believe those intermediate users can sufficiently use try-catch clauses to capture exceptions.